xla/service/cpu/onednn_ops_rewriter.cc

```diff
@@ -61,11 +63,35 @@ HloInstruction* FindLayerNormShift(HloInstruction* instr) {
   return shift;
 }
 
-std::optional<HloInstruction*> MatchSoftmax(HloInstruction* instr) {
+bool is_neg_inf_const_scalar(const HloInstruction* const_instr) {
+  if (!ShapeUtil::IsEffectiveScalar(const_instr->shape())) return false;
+  auto value = LiteralUtil::GetFirstScalarLiteral(const_instr->literal());
+  return literal_comparison::Equal(
+             value, LiteralUtil::MinValue(const_instr->shape().element_type()))
+      .ok();
+}
+
+auto MaxReduce(HloInstruction** instr) {
+  auto is_valid_reduce_max = [](const HloInstruction* reduce) {
+    HloComputation* reducer = reduce->to_apply();
+    return (reducer->root_instruction()->opcode() == HloOpcode::kMaximum) &&
+           (reduce->dimensions().size() == 1) &&
+           (reduce->operand(1)->opcode() == HloOpcode::kConstant) &&
+           is_neg_inf_const_scalar(reduce->operand(1));
+  };
+
+  return m::AnyOf<HloInstruction>(
+      m::Maximum().WithBinaryOperandsAnyOrder(
+          m::Reduce(instr).WithPredicate(is_valid_reduce_max).WithOneUse(),
+          m::Constant().WithPredicate(is_neg_inf_const_scalar)),
+      m::Reduce(instr).WithPredicate(is_valid_reduce_max).WithOneUse());
+}
+
+std::optional<HloInstruction*> MatchSoftmax(HloInstruction* instr, int* axis) {
```

Comment:
Can you clarify in the documentation what `instr` is supposed to correspond to? Is it the root of the operation?

Please also note that what we return is the `producer` of the pattern, and that the reduction axis is populated as a side effect on successful matches.

Reply:
Done

Reply:
Thank you this is clearer!